### DISCUSSION TITLE
AvaloniaRuntimeXamlLoader.Load - designMode?

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#6864] [Q&A] [answered]
Hi,

A couple, well 3, questions about the following call in the AvaloniaRuntimeXamlLoader class:

        public static class AvaloniaRuntimeXamlLoader

        // Summary:
        //     Loads XAML from a string.
        //
        // Parameters:
        //   xaml:
        //     The string containing the XAML.
        //
        //   localAssembly:
        //     Default assembly for clr-namespace:.
        //
        //   rootInstance:
        //     The optional instance into which the XAML should be loaded.
        //
        //   uri:
        //     The URI of the XAML being loaded.
        //
        //   designMode:
        //     Indicates whether the XAML is being loaded in design mode.
        //
        // Returns:
        //     The loaded object.
        public static object Load(string xaml, Assembly localAssembly = null, object rootInstance = null, Uri uri = null, bool designMode = false);

Q1. What it is the effect of passing designMode= true?

Q2. Is it possible to determine whether design mode is active from within a Window or UserControl class?

Q3. If the xml content is passed as a string, what purpose does the "uri" value serve?

Thanks

**Comment by kekekeks** [comment, marked as answer]
A1. If designMode is false `d:***` and `Design.***` properties are stripped. It has no other effects.
A2. "full" design mode is a process-wide setting, it's not possible to set it for individual controls
A3. uri is required for resolving relative resource paths

**Comment by kuiperzone** [comment]
Thank you

**Reply by kekekeks** [reply]
If you are interested in running the actual "design mode" (which is currently just live previewer), take a look at https://github.com/AvaloniaUI/Avalonia/wiki/XAML-previewer-protocol

**Reply by kuiperzone** [reply]
Thanks again. Will be taking a look.

**Reply by kuiperzone** [reply]
Hi @kekekeks ,

I've been looking at the XAML-previewer and, although I'm developing primarily for LINUX, I've been looking especially as the AvaloniaVS extension for Visual Studio as this should (presumably) be an example of how it is done.

But first off, I am absolutely baffled by the AvaloniaVS code. When I run Visual Studio on Windows with the "Avalonia for Visual Studio" extension, the XAML preview I see is fully interactive preview, with buttons etc. I can click on. However, when I look at the AvaloniaVS source code, I can not see where this is generated.

There is a `PreviewerProcess` class which fires up the Designer.HostApp, but appears to generate only a bitmap:

    public BitmapSource Bitmap => _bitmap;

This by the way, appears to be a Windows only solution, but presumably I could build an Avalonia WriteableBitmap from the raw data received over the socket.

But it remains unclear to me how the extension is generating preview of the XAML which is interactive, rather than just a dumb bitmap. I am possibly missing something obvious here, something to do with HWND perhaps?

I note that it starts Designer.HostApp.dll without specifying the `--method` option, but I'm guessing it defaults to `avalonia-remote`.

Q. Assuming I chose to go with the previewer, I note that as a pre-requisit it states:

"Previewer requires the final app executable that has BuildAvaloniaApp in the class that contains the application entry point and some information required to load app dependencies."

I'm not sure what this means. Does it mean the author of the target application must insert code into the class containing the entry point in order for the previewer to work? If so, why doesn't the VS extension require this?

PS. And thanks for all your help recently on other questions.

**Reply by kuiperzone** [reply]
For reference, here is the PreviewerProcess class I refer to:

https://github.com/AvaloniaUI/AvaloniaVS/blob/master/AvaloniaVS/Services/PreviewerProcess.cs

Presumably the "Avalonia Visual Studio" extension == AvaloniaVS?

**Reply by jp2masa** [reply]
For pointer events it's here: https://github.com/AvaloniaUI/AvaloniaVS/blob/544261dbca0078003bad98da41eae8c13afc6ea9/AvaloniaVS/Views/AvaloniaPreviewer.xaml.cs#L97-L136.

**Reply by kekekeks** [reply]
>Does it mean the author of the target application must insert code into the class containing the entry point in order for the previewer to work? If so, why doesn't the VS extension require this?


`BuildAvaloniaApp` method is provided in our [application templates](https://github.com/AvaloniaUI/avalonia-dotnet-templates/blob/d3b2f21a9df09de11d2b75f15c616358fbc270a3/templates/csharp/app/Program.cs#L18-L21)

**Reply by kuiperzone** [reply]
> For pointer events it's here: https://github.com/AvaloniaUI/AvaloniaVS/blob/544261dbca0078003bad98da41eae8c13afc6ea9/AvaloniaVS/Views/AvaloniaPreviewer.xaml.cs#L97-L136.

OMG! I see what you're doing. Thanks!

**Comment by kuiperzone** [comment]
Guys,

I'm writing a previewer targeting LINUX. As above, I've been looking at the `PreviewerProcess` class which fires up the `Designer.HostApp`.

I suspect that this isn't going to work on LINUX X11?

In order to generate the control, the HostApp calls `DesignWindowLoader.DesignWindowLoader()` which wraps the control loading in `DesignerMode`:

    using (PlatformManager.DesignerMode())
    {
        ....
        var loaded = AvaloniaRuntimeXamlLoader.Load(stream, localAsm, null, baseUri, true);
    }

The effect of `DesignerMode` is to set a flag, which causes "embeddable" windows to be created.

However, on X11 we have this:

        class AvaloniaX11Platform...

        public IWindowImpl CreateEmbeddableWindow()
        {
            throw new NotSupportedException();
        }

Am I correct in thinking that design time preview is only possible on Windows? Or can I overcome this?

**Reply by kekekeks** [reply]
Window embedding "transport" is only available for Win32. For Linux you need to use either avalonia-remote (which will send you frames  over the socket, example usage [can be found in AvalonStudio](https://github.com/VitalElement/AvalonStudio/tree/develop/AvalonStudio/AvalonStudio.Languages.Xaml)) or html (which will provide you with an local http URL that will serve previewer in any modern webview, example usage can be found in [vscode plugin prototype](https://github.com/rstm-sf/view-avalonia-preview-vscode)).
